docroot/modules/humsci/hs_config_readonly/src/EventSubscriber/ConfigReadOnlyEventSubscriber.php

```diff
@@ -0,0 +1,222 @@
+  }
+
+  public function onFormAlter(ReadOnlyFormEvent $event) {
+    // Check if the form is a ConfigFormBase or a ConfigEntityListBuilder.
+    $form_object = $event->getFormState()->getFormObject();
+    $mark_form_read_only = $form_object instanceof ConfigFormBase;
+
+    if (!$mark_form_read_only) {
+      $mark_form_read_only = in_array($form_object->getFormId(), $this->readOnlyFormIds);
+    }
+
+    // Check if the form is an EntityFormInterface and entity is a config
+    // entity.
+    if (!$mark_form_read_only && $form_object instanceof EntityFormInterface) {
+      $entity = $form_object->getEntity();
+      $mark_form_read_only = $entity instanceof ConfigEntityInterface;
+    }
+
+    // Don't block particular patterns.
+    if ($mark_form_read_only && $form_object instanceof EntityFormInterface) {
+      $entity = $form_object->getEntity();
+      $name = $entity->getConfigDependencyName();
+
+      // Block config from a module.
+      $mark_form_read_only = $this->configIsEditable([$name]);
+
+      // Unless its defined as a whiteist pattern.
+      if ($this->matchesWhitelistPattern($name)) {
+        $mark_form_read_only = FALSE;
+      }
+    }
+
+    // Config forms.
+    if ($mark_form_read_only && $form_object instanceof ConfigFormBase) {
+      $editable_config = $this->getEditableConfigNames($form_object);
+      $mark_form_read_only = $this->configIsEditable($editable_config);
+    }
+
+    if ($mark_form_read_only) {
+      $event->markFormReadOnly();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    $events = [];
+    $events[ReadOnlyFormEvent::NAME][] = ['onFormAlter', 200];
+    return $events;
+  }
+
+  /**
+   * Get the editable configuration names.
+   *
+   * @param ConfigFormBase $form
+   *   The configuration form.
+   *
+   * @return array
+   *   An array of configuration object names that are editable if called in
+   *   conjunction with the trait's config() method.
+   *
+   * @throws \ReflectionException
+   */
+  protected function getEditableConfigNames(ConfigFormBase $form) {
+    // Use reflection to work around getEditableConfigNames() as protected.
+    // @todo Review in 9.x for API change.
+    // @see https://www.drupal.org/node/2095289
+    $reflection = new \ReflectionMethod(get_class($form), 'getEditableConfigNames');
+    $reflection->setAccessible(TRUE);
+    return $reflection->invoke($form);
+  }
+
+  /**
+   * Check if the given configuration name can be edited.
+   *
+   * @param array $config
+   *   Array of config names.
+   *
+   * @return bool
+   *   If it can be edited.
+   */
+  protected function configIsEditable(array $config) {
+    if ($config == array_filter($config, [$this, 'isEditableConfig'])) {
+      return FALSE;
+    }
+    return TRUE;
+  }
+
+  /**
+   * @param $name
+   *
+   * @return bool
+   */
+  protected function isEditableConfig($name) {
```

Comment:
Why do we have two checks for editable config? Yes, I know they are different but it is confusing to have both of these functions esp without a docblock comment outlining why.